Scheduler

Kobzol opened the pull request:
Přidal jsem kontrolu importů distributed kontextu a Scheduler, zatím WIP.
Na začátku se rozhodne pro nějakou velikost batchů, a pak vždy proiteruje všechny aktuální batche. Potom se podívá, jestli jich má více a jestli ještě má čas. Jestli má co dělat, tak se podívá na posledních 10 jobů, jak dlouho trvaly, porovná to s "cílovou" dobou (teď je to 2 minuty) a upraví podle toho velikost příštích batchů.

Pokud tam je 20 batchů, tak se jich udělá všech 20 a až pak se generují nové, což může zbytečně nechat workery čekat na ostatní. Na druhou stranu pokud bych např. po každém došlém jobu  generoval nový task do clusteru, tak by to ta komunikace mohla dost zbrzdit.

Nenašel jsem timeout u future ani u executoru, vypadá to, že je distributed neimplementuje.
Je tam timeout jen u funkce, která čeká na vyhotovení všech futures v poli (http://distributed.readthedocs.io/en/latest/_modules/distributed/client.html#wait), což by se dalo použít s tou současnou implementací, ale nemohli bychom pak reagovat na jednotlivé future.
- zatím tam není proporcionální zrychlení/zpomalení podle dobíhajícího času
  Čekám na code review x)

Kobzol committed:
FIX: added check for missing/erroneous distributed import

Kobzol committed:
FIX: removed pypy from Travis build

Kobzol committed:
ENH: added basis of job scheduler

Kobzol committed:
ENH: removed generator from scheduler and added job truncating

Kobzol committed:
FIX: fixed leftover future handling

Kobzol committed:
TEST: changed distributed tests to use LocalCluster

Kobzol committed:
ENH: changed qsub run script to use PyPy

spirali commented on src/qit/base/runtime/distributedcontext.py:
```diff
@@ -5,15 +5,21 @@
+    from distributed import Scheduler, Nanny as Worker, Executor, as_completed
+    from tornado.ioloop import IOLoop
+
+    DistributedImportError = None
```
lower_case for global variables

Kobzol replied:
Sorry I don't understand, should they be named with lower_case or not? NoValue in iterator is in CamelCase too.

Kobzol committed:
FIX: fixed global variable case

Kobzol committed:
ENH: script for reading cluster info

Kobzol committed:
FIX: renamed samples to groups in tests

Kobzol commented:
Fixed global variable case and renamed samples to groups in tests.
Added a simple script to get the info about a running cluster (worker cpu/memory, task distribution etc.). Rebased everything onto spirali/master.

Kobzol committed:
REF: changed qit to haydi
